tunnel/build.gradle

```diff
@@ -7,6 +7,7 @@ group groupName
 
 android {
     compileSdk 33
+    ndkVersion '21.4.7075529'
```

Comment:
Can you share the build failure logs for the error you saw? I think we'd prefer to upgrade if possible.

Reply:
The error in question started after updating AGP from version 7.2.2 to 7.3.1, the new version of AGP (Android Gradle Plugin) has NDK 23.1.7779620 as default. This NDK change caused the following error when compiling on Ubuntu 22.04:

`Execution failed for task ':tunnel:buildCMakeDebug[arm64-v8a]'.
> com.android.ide.common.process.ProcessException: ninja: Entering directory `/home/anderson/IdeaProjects/wireguard-android/tunnel/.cxx/Debug/4z95p2w6/arm64-v8a'
  [1/16] Linking C executable /home/anderson/IdeaProjects/wireguard-android/tunnel/build/intermediates/cxx/Debug/4z95p2w6/obj/arm64-v8a/libwg-quick.so
  FAILED: /home/anderson/IdeaProjects/wireguard-android/tunnel/build/intermediates/cxx/Debug/4z95p2w6/obj/arm64-v8a/libwg-quick.so 
  : && /home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=aarch64-none-linux-android21 --gcc-toolchain=/home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -fno-limit-debug-info -static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -Wl,--gc-sections  -fuse-ld=gold CMakeFiles/libwg-quick.so.dir/wireguard-tools/src/wg-quick/android.c.o CMakeFiles/libwg-quick.so.dir/ndk-compat/compat.c.o -o /home/anderson/IdeaProjects/wireguard-android/tunnel/build/intermediates/cxx/Debug/4z95p2w6/obj/arm64-v8a/libwg-quick.so  -ldl  -latomic -lm && :
  /usr/bin/ld.gold: erro fatal: /home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/21/crtbegin_dynamic.o: unsupported ELF machine number 183
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
  [2/16] Building wireguard-go
  FAILED: CMakeFiles/libwg-go.so 
  cd /home/anderson/IdeaProjects/wireguard-android/tunnel/tools/libwg-go && make ANDROID_ARCH_NAME=arm64 ANDROID_C_COMPILER=/home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang ANDROID_TOOLCHAIN_ROOT=/home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64 ANDROID_LLVM_TRIPLE=aarch64-none-linux-android21 ANDROID_SYSROOT= ANDROID_PACKAGE_NAME=com.wireguard.android.debug GRADLE_USER_HOME=/home/anderson/.gradle "CFLAGS=-g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -Wno-unused-command-line-argument" "LDFLAGS=-static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments  -fuse-ld=gold" DESTDIR=/home/anderson/IdeaProjects/wireguard-android/tunnel/build/intermediates/cxx/Debug/4z95p2w6/obj/arm64-v8a BUILDDIR=/home/anderson/IdeaProjects/wireguard-android/tunnel/build/intermediates/cxx/Debug/4z95p2w6/obj/arm64-v8a/../generated-src
  go build -tags linux -ldflags="-X golang.zx2c4.com/wireguard/ipc.socketDirectory=/data/data/com.wireguard.android.debug/cache/wireguard" -v -trimpath -o "/home/anderson/IdeaProjects/wireguard-android/tunnel/build/intermediates/cxx/Debug/4z95p2w6/obj/arm64-v8a/libwg-go.so" -buildmode c-shared
  runtime/cgo
  # runtime/cgo
  /usr/bin/ld.gold: erro fatal: /home/anderson/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21/crtbegin_dynamic.o: unsupported ELF machine number 183`

I tried to find a solution to the problem but I wasn't very successful.😔

Reply:
Thanks, I figured out the problem. We explicitly ask for the [ld.gold linker](https://github.com/WireGuard/wireguard-android/blob/f269fb175dd7bdf5e2fa785ed451c2c81a529400/tunnel/tools/CMakeLists.txt#L10) in our build scripts because of a bug with `ld.bfd` but newer versions of the NDK no longer ship either of them, only `ld.ldd`. This causes the compiler to look for `ld.gold` elsewhere, and it picks up the one from your host operating system which cannot link ARM64 code and thus fails the build. Your fix seems correct as a stopgap solution, I'll merge this with a comment added explaining why we're doing it.